### PR TITLE
Fix buffer handling of syslog and timestamp parsers

### PR DESCRIPTION
**Pull request by MrAnno**
Partial fix for #4105

TODO:
- [x] additional tests
- [x] news entry

### CODE REVIEW

File: lib/timeutils/scan-timestamp.c [hunk 1]
```diff
@@ -34,41 +34,43 @@ scan_day_abbrev(const gchar **buf, gint *left, gint *wday)
 {
   *wday = -1;
 
-  if (*left < 3)
+  gsize abbrev_length = 3;
```
Comment (alltilla) [inline lib/timeutils/scan-timestamp.c:37]: [optional]

Maybe mark this as `const`?
```c
  const gsize abbrev_length = 3;
```

File: lib/timeutils/scan-timestamp.c [hunk 2]
```diff
@@ -87,63 +89,65 @@ scan_month_abbrev(const gchar **buf, gint *left, gint *mon)
 {
   *mon = -1;
 
-  if (*left < 3)
+  gsize abbrev_length = 3;
```
Comment (alltilla) [inline lib/timeutils/scan-timestamp.c:92]: [optional]

Same about marking as `const`.

**Comment by alltilla** [comment]
Maybe this could get a `left &&`, too: https://github.com/syslog-ng/syslog-ng/blob/master/modules/syslogformat/syslog-format.c#L210

---

[optional]

A lot of the places check for `left == 0` now, but if something goes really really bad, it can become `-1`. Should we maybe add some `g_assert(left >= 0)` ?

**Review by alltilla — DISMISSED** [review]
Nice catches! Grats :)

File: modules/syslogformat/tests/test_syslog_format.c [hunk 3]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2022 One Identity
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/syslogformat/tests/test_syslog_format.c:2]: A TDD-style commit history is okay in my opinion (the only requirement is that everything should compile).

File: modules/syslogformat/tests/test_syslog_format.c [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  app_shutdown();
+  cfg_free(cfg);
+}
+
+TestSuite(syslog_format, .init = setup, .fini = teardown);
+
+Test(syslog_format, parser_should_not_spin_on_non_zero_terminated_input, .timeout = 10)
+{
+  const gchar *data = "<182>2022-08-17T05:02:28.217 mymachine su: 'su root' failed for lonvick on /dev/pts/8";
+  gsize data_length = 27;
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/syslogformat/tests/test_syslog_format.c:61]: There is no point, I can remove it.

**Comment by MrAnno** [comment]
> Maybe this could get a left &&, too: https://github.com/syslog-ng/syslog-ng/blob/master/modules/syslogformat/syslog-format.c#L210

Nice catch, I'm fixing that too.

> Should we maybe add some g_assert(left >= 0) ?

We can add a few asserts to catch future bugs easily. The only aspect we need to be careful about is performance.

File: modules/syslogformat/tests/test_syslog_format.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/syslogformat/tests/test_syslog_format.c:61]: Sorry, I misunderstood the question first.
`27` was chosen carefully to reproduce one of the spinning behaviors.

[the remark this answers is not in the document]
Comment (alltilla) [reply modules/syslogformat/tests/test_syslog_format.c:61]: Can you indicate somehow, why it has to be `27` for the problem to happen? Comment or variable name would be nice.

[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/syslogformat/tests/test_syslog_format.c:63]: I've added a comment.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by MrAnno** [comment]
Oh, I've tried using new Criterion features, but that's not available in Kira. (The new assertion API shows git-like diffs on assert failures, which is excellent.)
I'll revert back to the old format.

**Review by alltilla — APPROVED** [review]
Good job!

**Review by Kokan — APPROVED** [review]
Executed the libfuzzer branch of mine for a while, it did not find new issue on this branch. Approve from me.

**Pull request merged**